pinot-common/src/main/java/org/apache/pinot/common/utils/helix/LeadControllerUtils.java

```diff
@@ -0,0 +1,97 @@
+import java.util.Set;
+import org.apache.helix.model.ExternalView;
+import org.apache.helix.model.MasterSlaveSMD;
+import org.apache.pinot.common.utils.HashUtils;
+import org.apache.pinot.common.utils.StringUtil;
+
+
+public class LeadControllerUtils {
+
+  /**
+   * Gets hash code for table using murmur2 function, ignores the most significant bit.
+   * Note: This method CANNOT be changed when lead controller resource is enabled.
+   * Otherwise it will assign different controller for the same table, which will mess up the controller periodic tasks and realtime segment completion.
+   * @param rawTableName table name
+   * @return hash code ignoring the most significant bit.
+   */
+  private static int getHashCodeForTable(String rawTableName) {
+    return (HashUtils.murmur2(StringUtil.encodeUtf8(rawTableName)) & 0x7fffffff);
+  }
+
+  /**
+   * Given a raw table name and number of partitions, returns the partition id in lead controller resource.
+   * @param rawTableName raw table name
+   * @param numPartitions number of partitions
+   * @return partition id in lead controller resource.
+   */
+  public static int getPartitionIdForTable(String rawTableName, int numPartitions) {
+    return getHashCodeForTable(rawTableName) % numPartitions;
+  }
+
+  /**
+   * Gets lead controller for table from lead controller resource.
+   * If the resource is disabled or no controller registered as participant, there is no instance in "MASTER" state.
+   * @param leadControllerResourceExternalView external view of lead controller resource
+   * @param rawTableName table name without type
+   * @return leader of partition, null if not found or resource is disabled.
+   */
+  public static String getLeadControllerForTable(ExternalView leadControllerResourceExternalView, String rawTableName) {
+    if (leadControllerResourceExternalView == null) {
+      return null;
+    }
+    Set<String> partitionSet = leadControllerResourceExternalView.getPartitionSet();
+    if (partitionSet == null || partitionSet.isEmpty()) {
+      return null;
+    }
+    int numPartitions = partitionSet.size();
+    int partitionIndex = getPartitionIdForTable(rawTableName, numPartitions);
+    String partitionName = formPartitionName(partitionIndex);
+    Map<String, String> partitionStateMap = leadControllerResourceExternalView.getStateMap(partitionName);
+
+    // Get master host from partition map. Return null if no master found.
+    for (Map.Entry<String, String> entry : partitionStateMap.entrySet()) {
+      if (MasterSlaveSMD.States.MASTER.name().equals(entry.getValue())) {
+        return entry.getKey();
+      }
+    }
+    return null;
+  }
+
+  public static String generateControllerParticipantId(String controllerHost, String controllerPort) {
+    return org.apache.pinot.common.utils.CommonConstants.Helix.PREFIX_OF_CONTROLLER_INSTANCE + controllerHost + "_"
+        + controllerPort;
+  }
+
+  public static String formPartitionName(int partitionIndex) {
```

Comment:
generatePartitionName() to keep the method naming consistent?

Reply:
Done.